File: GL_IOS/GL_IOS/Auth/GoogleWebOAuth.swift
import Foundation
import UIKit
import AuthenticationServices
import CryptoKit

// ════════════════════════════════════════════════════════════════════════════
// Google 로그인 — GoogleSignIn SDK 대신 ASWebAuthenticationSession 기반 웹 OAuth(PKCE).
// 받은 id_token/access_token 을 기존 Firebase signInWithGoogle(idToken, accessToken) 로 그대로 전달.
// 클라이언트 ID·리버스 스킴은 GoogleService-Info.plist 에서 로드(iOS OAuth 클라이언트, 시크릿 불필요).
// ════════════════════════════════════════════════════════════════════════════
final class GoogleWebOAuth: NSObject, ASWebAuthenticationPresentationContextProviding {
    static let shared = GoogleWebOAuth()
    private var session: ASWebAuthenticationSession?

    struct Tokens { let idToken: String; let accessToken: String?; let email: String?; let name: String?; let picture: String? }

    private var clientID: String { plist("CLIENT_ID") }
    private var reversedClientID: String { plist("REVERSED_CLIENT_ID") }
    private func plist(_ key: String) -> String {
        guard let path = Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist"),
              let dict = NSDictionary(contentsOfFile: path), let v = dict[key] as? String else { return "" }
        return v
    }

    /// 웹 OAuth 로그인 → (idToken, accessToken, email, name, picture). 실패 시 nil.
    func signIn(completion: @escaping (Tokens?) -> Void) {
        let verifier = Self.randomURLSafe(64)
        let challenge = Self.codeChallenge(verifier)
        let redirectURI = "\(reversedClientID):/oauth2redirect"

        var comp = URLComponents(string: "https://accounts.google.com/o/oauth2/v2/auth")!
        comp.queryItems = [
            .init(name: "client_id", value: clientID),
            .init(name: "redirect_uri", value: redirectURI),
            .init(name: "response_type", value: "code"),
            .init(name: "scope", value: "openid email profile"),
            .init(name: "code_challenge", value: challenge),
            .init(name: "code_challenge_method", value: "S256"),
            .init(name: "prompt", value: "select_account"),
        ]
        guard let authURL = comp.url else { completion(nil); return }

        let session = ASWebAuthenticationSession(url: authURL, callbackURLScheme: reversedClientID) { [weak self] callbackURL, error in
            guard let self else { completion(nil); return }
            guard let callbackURL,
                  let code = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false)?
                    .queryItems?.first(where: { $0.name == "code" })?.value else {
                if let error { NSLog("GoogleWebOAuth: \(error.localizedDescription)") }
                completion(nil); return
            }
            self.exchange(code: code, verifier: verifier, redirectURI: redirectURI, completion: completion)
        }
        session.presentationContextProvider = self
        session.prefersEphemeralWebBrowserSession = false
        self.session = session
        session.start()
    }

    /// authorization_code → 토큰 교환 (PKCE, iOS 클라이언트는 시크릿 없음)
    private func exchange(code: String, verifier: String, redirectURI: String, completion: @escaping (Tokens?) -> Void) {
        var req = URLRequest(url: URL(string: "https://oauth2.googleapis.com/token")!)
        req.httpMethod = "POST"
        req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        let form = [
            "client_id": clientID,
            "code": code,
            "code_verifier": verifier,
            "grant_type": "authorization_code",
            "redirect_uri": redirectURI,
        ]
        req.httpBody = form.map { "\($0.key)=\(Self.formEncode($0.value))" }.joined(separator: "&").data(using: .utf8)

        URLSession.shared.dataTask(with: req) { data, _, _ in
            guard let data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let idToken = json["id_token"] as? String else { completion(nil); return }
            let claims = Self.decodeJWT(idToken)
            completion(Tokens(
                idToken: idToken,
                accessToken: json["access_token"] as? String,
                email: claims["email"] as? String,
                name: claims["name"] as? String,
                picture: claims["picture"] as? String
            ))
        }.resume()
    }

    // ── PKCE / 인코딩 헬퍼 ──
    private static func randomURLSafe(_ n: Int) -> String {
        var bytes = [UInt8](repeating: 0, count: n)
        _ = SecRandomCopyBytes(kSecRandomDefault, n, &bytes)
        return Data(bytes).base64URLEncoded()
    }
    private static func codeChallenge(_ verifier: String) -> String {
        Data(SHA256.hash(data: Data(verifier.utf8))).base64URLEncoded()
    }
    private static func formEncode(_ s: String) -> String {
        var cs = CharacterSet.alphanumerics
        cs.insert(charactersIn: "-._~")
        return s.addingPercentEncoding(withAllowedCharacters: cs) ?? s
    }
    private static func decodeJWT(_ jwt: String) -> [String: Any] {
        let parts = jwt.split(separator: ".")
        guard parts.count >= 2 else { return [:] }
        var b64 = String(parts[1]).replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        while b64.count % 4 != 0 { b64 += "=" }
        guard let data = Data(base64Encoded: b64),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return [:] }
        return json
    }

    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        UIApplication.shared.connectedScenes
            .compactMap { ($0 as? UIWindowScene)?.keyWindow }.first ?? ASPresentationAnchor()
    }
}

private extension Data {
    func base64URLEncoded() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: Gatcha Log Android/src/main/java/com/gatcha/log/auth/GoogleWebOAuth.kt
package com.gatcha.log.auth

import android.content.Context
import android.net.Uri
import android.util.Base64
import android.util.Log
import androidx.browser.customtabs.CustomTabsIntent
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * Google 로그인 — 브라우저 기반 웹 OAuth(PKCE). 기존 Credential Manager(기기 구글 계정 필수) 대체.
 *
 * iOS OAuth 클라이언트 ID + 리버스 스킴을 재사용한다(시크릿 불필요·PKCE, Google Console 변경 불필요).
 * 받은 id_token 은 Firebase(GoogleAuthProvider) 인증에 그대로 전달. 동의 화면이 production 이면 모든 유저 로그인 가능.
 */
object GoogleWebOAuth {
    // GoogleService-Info(iOS) 의 CLIENT_ID / REVERSED_CLIENT_ID — 공개값(시크릿 아님).
    private const val CLIENT_ID = "711708512022-6upq632okh78jru6q09o2dol4tekokbn.apps.googleusercontent.com"
    private const val SCHEME = "com.googleusercontent.apps.711708512022-6upq632okh78jru6q09o2dol4tekokbn"
    private const val REDIRECT_URI = "$SCHEME:/oauth2redirect"

    data class Tokens(val idToken: String, val email: String?, val name: String?, val picture: String?)

    private var pending: CompletableDeferred<Uri?>? = null
    private var verifier: String = ""

    /** 리다이렉트 액티비티가 호출 — 콜백 URI 전달. */
    fun onRedirect(uri: Uri?) {
        pending?.complete(uri)
        pending = null
    }

    /** 브라우저 웹 로그인 → 토큰. 취소/실패 시 null. */
    suspend fun signIn(context: Context): Tokens? {
        pending?.complete(null) // 잔여 시도 정리
        verifier = randomUrlSafe(64)
        val challenge = b64url(sha256(verifier.toByteArray(Charsets.US_ASCII)))
        val authUrl = Uri.parse("https://accounts.google.com/o/oauth2/v2/auth").buildUpon()
            .appendQueryParameter("client_id", CLIENT_ID)
            .appendQueryParameter("redirect_uri", REDIRECT_URI)
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("scope", "openid email profile")
            .appendQueryParameter("code_challenge", challenge)
            .appendQueryParameter("code_challenge_method", "S256")
            .appendQueryParameter("prompt", "select_account")
            .build()

        val deferred = CompletableDeferred<Uri?>()
        pending = deferred
        runCatching {
            // 인앱 브라우저(Chrome Custom Tabs) — Activity 컨텍스트로 같은 태스크에 띄워 인앱 느낌.
            // Custom Tabs 미지원 브라우저면 launchUrl 이 일반 브라우저로 자동 폴백한다.
            CustomTabsIntent.Builder().setShowTitle(true).build().launchUrl(context, authUrl)
        }.onFailure { pending = null; return null }

        val callback = deferred.await() ?: return null
        val code = callback.getQueryParameter("code") ?: return null
        return withContext(Dispatchers.IO) { exchange(code) }
    }

    private fun exchange(code: String): Tokens? {
        val body = listOf(
            "client_id" to CLIENT_ID,
            "code" to code,
            "code_verifier" to verifier,
            "grant_type" to "authorization_code",
            "redirect_uri" to REDIRECT_URI,
        ).joinToString("&") { "${it.first}=${Uri.encode(it.second)}" }

        val conn = (URL("https://oauth2.googleapis.com/token").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            doOutput = true
            setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
        }
        return try {
            conn.outputStream.use { it.write(body.toByteArray()) }
            if (conn.responseCode !in 200..299) {
                Log.e("GoogleWebOAuth", "token exchange HTTP ${conn.responseCode}")
                return null
            }
            val json = JSONObject(conn.inputStream.bufferedReader().readText())
            val idToken = json.optString("id_token").ifEmpty { return null }
            val claims = decodeJwt(idToken)
            Tokens(
                idToken = idToken,
                email = claims.optString("email").ifEmpty { null },
                name = claims.optString("name").ifEmpty { null },
                picture = claims.optString("picture").ifEmpty { null },
            )
        } catch (e: Exception) {
            Log.e("GoogleWebOAuth", "token exchange failed", e)
            null
        } finally {
            conn.disconnect()
        }
    }

    private fun sha256(b: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(b)
    private fun b64url(b: ByteArray): String = Base64.encodeToString(b, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
    private fun randomUrlSafe(n: Int): String = b64url(ByteArray(n).also { SecureRandom().nextBytes(it) })
    private fun decodeJwt(jwt: String): JSONObject {
        val parts = jwt.split(".")
        if (parts.size < 2) return JSONObject()
        return try {
            JSONObject(String(Base64.decode(parts[1], Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP), Charsets.UTF_8))
        } catch (e: Exception) {
            JSONObject()
        }
    }
}
